homeassistant/components/tado/coordinator.py

```diff
@@ -0,0 +1,433 @@
+    PRESET_AUTO,
+    TEMP_OFFSET,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=4)
+SCAN_INTERVAL = timedelta(minutes=5)
+SCAN_MOBILE_DEVICE_INTERVAL = timedelta(seconds=30)
+
+type TadoConfigEntry = ConfigEntry[TadoDataUpdateCoordinator]
+
+
+class TadoDataUpdateCoordinator(DataUpdateCoordinator[dict[str, dict]]):
+    """Class to manage API calls from and to Tado via PyTado."""
+
+    tado: Tado
+    home_id: int
+    home_name: str
+    config_entry: TadoConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        tado: Tado,
+        debug: bool = False,
+    ) -> None:
+        """Initialize the Tado data update coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+        self._tado = tado
+        self._username = entry.data[CONF_USERNAME]
+        self._password = entry.data[CONF_PASSWORD]
+        self._fallback = entry.options.get(CONF_FALLBACK, CONST_OVERLAY_TADO_DEFAULT)
+        self._debug = debug
+
+        self.home_id: int
+        self.home_name: str
+        self.zones: list[dict[Any, Any]] = []
+        self.devices: list[dict[Any, Any]] = []
+        self.data: dict[str, dict] = {
+            "device": {},
+            "mobile_device": {},
+            "weather": {},
+            "geofence": {},
+            "zone": {},
+        }
+
+    @property
+    def fallback(self) -> str:
+        """Return fallback flag to Smart Schedule."""
+        return self._fallback
+
+    async def _async_setup(self) -> None:
+        """Set up and load initial data."""
+
+        try:
+            _LOGGER.debug("Preloading home data")
+            tado_home_call = await self.hass.async_add_executor_job(self._tado.get_me)
+            _LOGGER.debug("Preloading zones and devices")
+            self.zones = await self.hass.async_add_executor_job(self._tado.get_zones)
+            self.devices = await self.hass.async_add_executor_job(
+                self._tado.get_devices
+            )
+        except RequestException as err:
+            raise UpdateFailed(f"Error during Tado setup: {err}") from err
+
+        tado_home = tado_home_call["homes"][0]
+        self.home_id = tado_home["id"]
+        self.home_name = tado_home["name"]
+        _LOGGER.info("Tado setup complete")
+
+    async def _async_update_data(self) -> dict[str, dict]:
+        """Fetch the latest data from Tado."""
+
+        try:
+            # Fetch updated data for devices, mobile devices, zones, and home
+            devices = await self._async_update_devices()
+            mobile_devices = await self._async_update_mobile_devices()
+            zones = await self._async_update_zones()
+            home = await self._async_update_home()
+        except RequestException as err:
+            raise UpdateFailed(f"Error updating Tado data: {err}") from err
```

Comment:
Also, you're already catching a lot of `RequestException` in the functions and reraising them as `UpdateFailed`